deps/flatgeobuf/packedrtree.cpp

```diff
@@ -0,0 +1,410 @@
+    i1 = (i1 | (i1 << 4)) & 0x0F0F0F0F;
+    i1 = (i1 | (i1 << 2)) & 0x33333333;
+    i1 = (i1 | (i1 << 1)) & 0x55555555;
+
+    uint32_t value = ((i1 << 1) | i0);
+
+    return value;
+}
+
+uint32_t hilbert(const NodeItem &r, uint32_t hilbertMax, const double minX, const double minY, const double width, const double height)
+{
+    uint32_t x = 0;
+    uint32_t y = 0;
+    uint32_t v;
+    if (width != 0.0)
+        x = static_cast<uint32_t>(floor(hilbertMax * ((r.minX + r.maxX) / 2 - minX) / width));
+    if (height != 0.0)
+        y = static_cast<uint32_t>(floor(hilbertMax * ((r.minY + r.maxY) / 2 - minY) / height));
+    v = hilbert(x, y);
+    return v;
+}
+
+const uint32_t hilbertMax = (1 << 16) - 1;
+
+void hilbertSort(std::vector<std::shared_ptr<Item>> &items)
+{
+    NodeItem extent = calcExtent(items);
+    const double minX = extent.minX;
+    const double minY = extent.minY;
+    const double width = extent.width();
+    const double height = extent.height();
+    std::sort(items.begin(), items.end(), [minX, minY, width, height] (std::shared_ptr<Item> a, std::shared_ptr<Item> b) {
+        uint32_t ha = hilbert(a->nodeItem, hilbertMax, minX, minY, width, height);
+        uint32_t hb = hilbert(b->nodeItem, hilbertMax, minX, minY, width, height);
+        return ha > hb;
+    });
+}
+
+void hilbertSort(std::vector<NodeItem> &items)
+{
+    NodeItem extent = calcExtent(items);
+    const double minX = extent.minX;
+    const double minY = extent.minY;
+    const double width = extent.width();
+    const double height = extent.height();
+    std::sort(items.begin(), items.end(), [minX, minY, width, height] (const NodeItem &a, const NodeItem &b) {
+        uint32_t ha = hilbert(a, hilbertMax, minX, minY, width, height);
+        uint32_t hb = hilbert(b, hilbertMax, minX, minY, width, height);
+        return ha > hb;
+    });
+}
+
+NodeItem calcExtent(const std::vector<std::shared_ptr<Item>> &items)
+{
+    return std::accumulate(items.begin(), items.end(), NodeItem::create(0), [] (NodeItem a, const std::shared_ptr<Item>& b) {
+        return a.expand(b->nodeItem);
+    });
+}
+
+NodeItem calcExtent(const std::vector<NodeItem> &nodes)
+{
+    return std::accumulate(nodes.begin(), nodes.end(), NodeItem::create(0), [] (NodeItem a, const NodeItem &b) {
+        return a.expand(b);
+    });
+}
+
+void PackedRTree::init(const uint16_t nodeSize)
+{
+    if (nodeSize < 2)
+        throw std::invalid_argument("Node size must be at least 2");
+    if (_numItems == 0)
+        throw std::invalid_argument("Cannot create empty tree");
+    _nodeSize = std::min(std::max(nodeSize, static_cast<uint16_t>(2)), static_cast<uint16_t>(65535));
+    _levelBounds = generateLevelBounds(_numItems, _nodeSize);
+    _numNodes = _levelBounds.front().second;
+    _nodeItems = new NodeItem[static_cast<size_t>(_numNodes)];
+}
+
+std::vector<std::pair<uint64_t, uint64_t>> PackedRTree::generateLevelBounds(const uint64_t numItems, const uint16_t nodeSize) {
```

Comment:
This looks exactly like new gist build logic. @x4m and I have a theory that whole tree will be a lot better if we run picksplit on consecutive pairs of pages, deoverlapping them.

https://github.com/postgis/postgis/blob/f6def67654c25d812446239036cee44812613748/postgis/gserialized_spgist_2d.c#L337

Reply:
Out of my depth :) This is based on https://github.com/mourner/flatbush (but up side down)